server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -442,17 +448,6 @@ public void applyClusterState(ClusterChangedEvent event) {
                     }
                 }
             }
-            // Cancel cancellable tasks for the nodes that are gone
```

Comment:
I wasn't expecting this to be included here, although perhaps it makes sense as a separate change. It's reasonable if the node is shut down but today we also cancel these tasks if the node is temporarily partitioned from the master, and perhaps we want to keep that behaviour.

The issue is that we don't tell the partitioned node that it has been dropped from the cluster so it won't close its outgoing channels, which means we won't cancel those tasks any more if we remove this.

Also looks like the tests around this are quite weak, since this change doesn't seem to have a corresponding change to the tests.

Reply:
This has become an issue for us in the context of cross clusters. A subtask on the remote cluster will get canceled if any node in that cluster leaves. The reason is that the node with the parent task is from another cluster. This blocks the work to support cancellation cross clusters in https://github.com/elastic/elasticsearch/pull/55779. We would like to replace it with the [task heartbeat](https://github.com/elastic/elasticsearch/pull/56619), but I found your proposal is much better.

> Also looks like the tests around this are quite weak, since this change doesn't seem to have a corresponding change to the tests.

Yeah, we only test when a node [leaves](https://github.com/elastic/elasticsearch/blob/6ef9506bebd17a1a6fe2b97ef1b7721704d7d9b2/server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksTests.java#L372) the cluster.

Reply:
I see, in which case this change is ok with me. Could we make that test stronger? It seems very artificial that the tasks are only cancelled on `testNodes[0].close();`.

Reply:
> The issue is that we don't tell the partitioned node that it has been dropped from the cluster so it won't close its outgoing channels, which means we won't cancel those tasks any more if we remove this.

That looks like an enhancement to me. If the channel is still active we should give a chance for the task to finish ?

Reply:
ok, I pushed https://github.com/elastic/elasticsearch/pull/56620/commits/8ae70421eaffc6fc67f4fcaf74030099c9c01668.